mobile_verifier/src/reward_share.rs

```diff
@@ -0,0 +1,297 @@
+}
+
+impl GatheredShares {
+    pub async fn from_stream(
+        stream: &mut BytesMutStream,
+        after_utc: u64,
+        before_utc: u64,
+    ) -> Result<Self> {
+        let mut shares = Self {
+            shares: Shares::new(),
+            invalid_shares: InvalidShares::new(),
+            speed_shares: SpeedShares::new(),
+            speed_shares_moving_avg: SpeedShareMovingAvgs::default(),
+            invalid_speed_shares: InvalidSpeedShares::new(),
+        };
+
+        while let Some(Ok(msg)) = stream.next().await {
+            // NOTE: This will early exit with an error if we fail to decode
+
+            if let Ok(cell_heartbeat_req_v1) = CellHeartbeatReqV1::decode(msg.clone()) {
+                shares.gather_heartbeat(cell_heartbeat_req_v1, after_utc, before_utc)
+            } else if let Ok(speedtest_req_v1) = SpeedtestReqV1::decode(msg) {
+                shares.gather_speedtest(speedtest_req_v1, after_utc, before_utc)
+            } else {
+                continue;
+            }
+        }
+        Ok(shares)
+    }
+
+    fn gather_speedtest(&mut self, speedtest_req: SpeedtestReqV1, after_utc: u64, before_utc: u64) {
+        // TODO: Better error handling for bad public keys
+        let mut speed_share = match SpeedShare::try_from(speedtest_req) {
+            Ok(share) => share,
+            Err(_) => return,
+        };
+
+        if speed_share.timestamp < after_utc || speed_share.timestamp >= before_utc {
+            speed_share.validity = ShareValidity::HeartbeatOutsideRange;
+            self.invalid_speed_shares.push(speed_share);
+        } else {
+            self.speed_shares
+                .entry(speed_share.pub_key.clone())
+                .or_default()
+                .push(speed_share);
+            self.speed_shares_moving_avg.update(&self.speed_shares)
+        }
+    }
+
+    fn gather_heartbeat(
+        &mut self,
+        cell_heartbeat_req_v1: CellHeartbeatReqV1,
+        after_utc: u64,
+        before_utc: u64,
+    ) {
+        let CellHeartbeatReqV1 {
+            pub_key: hb_pub_key,
+            cbsd_id: hb_cbsd_id,
+            timestamp: hb_timestamp,
+            ..
+        } = cell_heartbeat_req_v1;
+
+        if let Some(cell_type) = CellType::from_cbsd_id(&hb_cbsd_id) {
+            // TODO: Will only get inserted in invalid if cbsd_id is a valid cell_type
+            if hb_timestamp < after_utc || hb_timestamp >= before_utc {
+                self.invalid_shares.push(ShareProto {
+                    cbsd_id: hb_cbsd_id,
+                    timestamp: hb_timestamp,
+                    pub_key: hb_pub_key.to_vec(),
+                    weight: crate::bones_to_u64(cell_type.reward_weight()),
+                    cell_type: cell_type as i32,
+                    validity: ShareValidity::HeartbeatOutsideRange as i32,
+                });
+                return;
+            }
+
+            if let Ok(gw_pubkey) = PublicKey::try_from(hb_pub_key.as_ref()) {
```

Comment:
```suggestion
            if let Ok(gw_pubkey) = PublicKey::try_from(hb_pub_key.as_slice()) {
```

I've started using this